rescpu.c

```diff
@@ -0,0 +1,196 @@
+			cpu->processor = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "vendor_id"))
+			strcpy(cpu->vendor_id, end2);
+		if (!strcmp(start, "cpu family"))
+			strcpy(cpu->cpu_family, end2);
+		if (!strcmp(start, "model"))
+			strcpy(cpu->model, end2);
+		if (!strcmp(start, "model name"))
+			strcpy(cpu->model_name, end2);
+		if (!strcmp(start, "stepping"))
+			strcpy(cpu->stepping, end2);
+		if (!strcmp(start, "microcode"))
+			strcpy(cpu->microcode, end2);
+		if (!strcmp(start, "cpu MHz"))
+			strcpy(cpu->cpu_mhz, end2);
+		if (!strcmp(start, "cache size"))
+			strcpy(cpu->cache_size, end2);
+		if (!strcmp(start, "physical id"))
+			cpu->physical_id = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "siblings"))
+			cpu->siblings = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "core id"))
+			cpu->core_id = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "cpu cores"))
+			cpu->cpu_cores = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "apicid"))
+			cpu->apicid = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "initial apicid"))
+			cpu->initial_apicid = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "fpu"))
+			strcpy(cpu->fpu, end2);
+		if (!strcmp(start, "fpu_exception"))
+			strcpy(cpu->fpu_exception, end2);
+		if (!strcmp(start, "cpuid level"))
+			cpu->cpu_id_level = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "wp"))
+			strcpy(cpu->wp, end2);
+		if (!strcmp(start, "flags"))
+			strcpy(cpu->flags, end2);
+		if (!strcmp(start, "vmx flags"))
+			strcpy(cpu->vmx_flags, end2);
+		if (!strcmp(start, "bugs"))
+			strcpy(cpu->bugs, end2);
+		if (!strcmp(start, "bogomips"))
+			cpu->bogomips = atof(end2);
+		if (!strcmp(start, "clflush size"))
+			cpu->clflush_size = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "cache_alignment"))
+			cpu->cache_alignment = strtoul(end2, NULL, 10);
+		if (!strcmp(start, "address sizes"))
+			strcpy(cpu->address_sizes, end2);
+		if (!strcmp(start, "power management"))
+			strcpy(cpu->power_mgmt, end2);
+
+nextline:
+		start = start2;
+		start++;
+		bytes_done++;
+	}
+	return 0;
+}
+
+static inline int get_buffer(char *fname, char *buf, unsigned int bufsz)
+{
+	FILE *fp;
+	int count;
+
+	fp = fopen(fname, "r");
+	if (fp == NULL) {
+		eprintf("Error opening File %s with errno: %d",
+			fname, errno);
+		return -1;
+	}
+	count = fread(buf, sizeof(char), bufsz-1, fp);
+	if (count < 0) {
+		eprintf("in read from File %s with errno: %d", fname, errno);
+		fclose(fp);
+		return -1;
+	}
+	fclose(fp);
+	buf[count] = '\0';
+	return count;
+}
+
+int getcpuinfo(int res_id, void *out, size_t sz, void **hint, int flags)
+{
+	int len, cpu_num;
+	char *buffer = NULL;
+
+	cpu_num = sysconf(_SC_NPROCESSORS_ONLN);
+
+	switch (res_id) {
+	case RES_CPU_INFO:
+		CHECK_SIZE(sz, (cpu_num * sizeof(struct cpuinfo)));
+		buffer = (char *) malloc(cpu_num * sizeof(struct cpuinfo));
+		if (buffer == NULL) {
+			errno = ENOMEM;
+			return -1;
+		}
+#ifdef TESTING
+		len = get_buffer("./cpu_info.orig", buffer, sz);
```

Comment:
Two questions regarding the testing:
1. I don't see Makefile compiling with -DTESTING
2. I don't see anything writing out a good cpu_info.orig, or any samples/ directory from which tests would read known-good info.

It would be good to use a github action in .github/workflow/build.yml which writes good files, builds the test, and checks the test output.